src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,513 @@
+            // We also need to only do this for the lines that are part of the input range to respect range formatting.
+            var desiredIndentationLevel = context.Indentations[(int)codeBlockRange.Start.Line].IndentationLevel + 1;
+            var editsToApply = new List<TextChange>();
+            var inputSpan = context.Range.AsTextSpan(originalText);
+            TrackChangeInSpan(originalText, inputSpan, changedText, out var changedInputSpan, out _);
+            var changedInputRange = changedInputSpan.AsRange(changedText);
+
+            for (var i = (int)changedInputRange.Start.Line; i <= changedInputRange.End.Line; i++)
+            {
+                var line = changedText.Lines[i];
+                if (line.Span.Length == 0)
+                {
+                    // Empty line. C# formatter didn't remove it so we won't either.
+                    continue;
+                }
+
+                if (!changedCodeBlockSpan.Contains(line.Start))
+                {
+                    // Defensive check to make sure we're not handling lines that are not part of the current code block.
+                    continue;
+                }
+
+                var leadingWhitespace = line.GetLeadingWhitespace();
+                var minCSharpIndentLength = context.Options.TabSize * minCSharpIndentLevel;
+                if (leadingWhitespace.Length < minCSharpIndentLength)
+                {
+                    // For whatever reason, the C# formatter decided to not indent this. Leave it as is.
+                    continue;
+                }
+                else
+                {
+                    var effectiveDesiredIndentationLevel = desiredIndentationLevel - minCSharpIndentLevel;
+                    if (effectiveDesiredIndentationLevel < 0)
+                    {
+                        // This means that we need to unindent.
+                        var length = Math.Abs(effectiveDesiredIndentationLevel * context.Options.TabSize);
+                        var span = new TextSpan(line.Start, (int)length);
+                        editsToApply.Add(new TextChange(span, string.Empty));
+                    }
+                    else if (effectiveDesiredIndentationLevel > 0)
+                    {
+                        // This means that we need to indent.
+                        var effectiveDesiredIndentation = GetIndentationString(context, effectiveDesiredIndentationLevel);
+                        var span = new TextSpan(line.Start, 0);
+                        editsToApply.Add(new TextChange(span, effectiveDesiredIndentation));
+                    }
+                }
+            }
+
+            changedText = ApplyChanges(changedText, editsToApply);
+            return changedText;
+        }
+
+        private void FormatCodeBlockStart(FormattingContext context, SourceText changedText, RazorDirectiveBodySyntax directiveBody, SyntaxNode innerCodeBlock, List<TextEdit> edits)
```

Comment:
`innerCodeBlock` vs. `directiveBody` what scenario results in them encompassing different ranges?

Reply:
None. I can technically obtain the `innerCodeBlock` from the `directiveBody`. Given this is just a private method, I just passed it in directly because I already had it.